sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -222,6 +222,20 @@ abstract class Expression extends TreeNode[Expression] {
    */
   def childrenResolved: Boolean = children.forall(_.resolved)
 
+  // Return a roughly canonicalized expression, which recursively canonicalizes each expression node
+  // without changing the structure of the expression tree.
+  // This method should only be called in `canonicalized` or subclass's `roughlyCanonicalized`.
+  // Subclasses (especially leaf nodes) should override this method to customize the
+  // canonicalization logic. Some basic canonicalization rules:
+  //  - Names and nullability hints for `DataType`s are stripped. See the override in
+  //    `AttributeReference` and `GetStructField`.
+  //  - TimeZoneId for [[Cast]] and [[AnsiCast]] are stripped if `needsTimeZone` is false.
+  //  - Rearrange the inputs of certain predicate expressions according to the `hashCode` of inputs.
+  lazy val roughlyCanonicalized: Expression = {
```

Comment:
`preCanonicalized`? roughlyCanonicalized` looks a bit vague and weird.

Reply:
Yeah .. or `fastCanonicalized`?